test/core/util/test_lb_policies.cc

```diff
@@ -746,6 +748,173 @@ class FailLbFactory : public LoadBalancingPolicyFactory {
+                RefCountedPtr<QueueOnceLoadBalancingPolicy>(this)),
+            std::move(args), "pick_first",
+            /*initial_refcount=*/2) {}
+
+  ~QueueOnceLoadBalancingPolicy() override = default;
+
+  absl::string_view name() const override { return kQueueOncePolicyName; }
+
+ private:
+  class Helper;
+
+  // An always queueing picker
+  class QueueingPicker : public SubchannelPicker {
+   public:
+    QueueingPicker(Helper* helper) : helper_(helper) {}
+
+    PickResult Pick(PickArgs /*args*/) override;
+
+   private:
+    Helper* helper_;
+  };
+
+  class DelegatingPicker : public SubchannelPicker {
+   public:
+    DelegatingPicker(RefCountedPtr<SubchannelPicker> delegate_picker)
+        : delegate_picker_(std::move(delegate_picker)) {}
+
+    PickResult Pick(PickArgs args) override {
+      return delegate_picker_->Pick(args);
+    }
+
+   private:
+    RefCountedPtr<SubchannelPicker> delegate_picker_;
+  };
+
+  class Helper : public ChannelControlHelper {
+   public:
+    Helper(RefCountedPtr<QueueOnceLoadBalancingPolicy> parent)
+        : parent_(std::move(parent)) {}
+
+    RefCountedPtr<SubchannelInterface> CreateSubchannel(
+        ServerAddress address, const ChannelArgs& args) override {
+      return parent_->channel_control_helper()->CreateSubchannel(
+          std::move(address), args);
+    }
+
+    void UpdateState(grpc_connectivity_state state, const absl::Status& status,
+                     RefCountedPtr<SubchannelPicker> picker) override {
+      bool pick_queued;
+      {
+        MutexLock lock(&mu_);
+        pick_queued = pick_queued_;
+        // If a pick has not been queued yet, we need to save the update from
+        // PickFirst to propagate later.
+        if (!pick_queued) {
+          state_to_update_ = {state, status, std::move(picker)};
+        }
+      }
+      if (!pick_queued) {
+        return parent_->channel_control_helper()->UpdateState(
+            GRPC_CHANNEL_CONNECTING, absl::OkStatus(),
+            MakeRefCounted<QueueingPicker>(this));
+      }
+      parent_->channel_control_helper()->UpdateState(
+          state, status, MakeRefCounted<DelegatingPicker>(std::move(picker)));
+    }
+
+    void RequestReresolution() override {
+      parent_->channel_control_helper()->RequestReresolution();
+    }
+
+    absl::string_view GetAuthority() override {
+      return parent_->channel_control_helper()->GetAuthority();
+    }
+
+    grpc_event_engine::experimental::EventEngine* GetEventEngine() override {
+      return parent_->channel_control_helper()->GetEventEngine();
+    }
+
+    void AddTraceEvent(TraceSeverity severity,
+                       absl::string_view message) override {
+      parent_->channel_control_helper()->AddTraceEvent(severity, message);
+    }
+
+    void SetPickQueued() {
+      MutexLock lock(&mu_);
+      pick_queued_ = true;
+      if (state_to_update_.has_value()) {
+        // Propagate the update from PickFirst but do not do it inline to avoid
```

Comment:
I don't think there's any need to worry about sequencing here, since we'll be caching the update from the child, so we'll always return the most recent picker.

Reply:
Right, I was actually just trying to synchronize access to `state_to_update_`. With the new logic of popping into the WorkSerializer from the picker, we just use the WorkSerializer to synchronize that access, and it all works.